services/graph/pkg/identity/ldap.go

```diff
@@ -1057,3 +1029,32 @@ func (i *LDAP) usersEnabledState(users []*ldap.Entry) (usersEnabledState map[str
 
 	return usersEnabledState, nil
 }
+
+// Behavior of enabling/disabling of users depends on the "disableUserMechanism" config option:
+//
+// "attribute": For the upstream user management service which modifies accountEnabled on the user entry
+// "group": Makes it possible for local admins to disable users by adding them to a special group
+func (i *LDAP) updateAccountEnabledState(logger log.Logger, accountEnabled bool, e *ldap.Entry, mr *ldap.ModifyRequest) (updateNeeded bool, err error) {
+	switch i.disableUserMechanism {
+	case DisableMechanismNone:
+		err = errors.New("accountEnabled option not compatible with backend disable user mechanism")
+	case DisableMechanismAttribute:
+		boolString := strings.ToUpper(strconv.FormatBool(accountEnabled))
+		ldapValue := e.GetEqualFoldAttributeValue(i.userAttributeMap.accountEnabled)
+		if ldapValue != "" {
+			mr.Replace(i.userAttributeMap.accountEnabled, []string{boolString})
+		} else {
+			mr.Add(i.userAttributeMap.accountEnabled, []string{boolString})
+		}
+		updateNeeded = true
+	case DisableMechanismGroup:
+		if accountEnabled {
+			err = i.enableUser(logger, e.DN)
+		} else {
+			err = i.disableUser(logger, e.DN)
+		}
+		updateNeeded = true
+	}
+
+	return updateNeeded, err
+}
```

Comment:
This switch case makes me think that an `i.disableMechanism func(mr *ldap.ModifyRequest, e *ldap.Entry, enabled bool) error` with three different implementations (DisableMechanismNoop, DisableMechanismAttribute and DisableMechanismGroup) that are chosen / configured at initialization would be better to read and extend.

Could you refactor this?

Reply:
@butonic I don't quite follow what you want me to do. Do you want me to store the function to use on initialization?

Reply:
Discussed with Ralf and we came to the conclusion that this can be done later.